alg/los.cpp

```diff
@@ -0,0 +1,212 @@
+// https://gist.github.com/ssavi-ict/092501c69e2ffec65e96a8865470ad2f
+// https://blog.demofox.org/2015/01/17/bresenhams-drawing-algorithms/
+// https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm
+// https://www.cs.helsinki.fi/group/goa/mallinnus/lines/bresenh.html
+// https://stackoverflow.com/questions/10060046/drawing-lines-with-bresenhams-line-algorithm
+// http://www.edepot.com/linebresenham.html
+// 3D:
+// https://gist.github.com/yamamushi/5823518
+
+// Run bresenham terrain checking from (x1, y1) to (x2, y2).
+// The callback is run at every point along the line,
+// which should return True if the point is above terrain.
+// Bresenham2D will return true if all points have LOS between the start and end.
+static
+bool Bresenham2D(int x1, int y1, int x2, int y2, std::function< auto(int, int) -> bool> OnBresenhamPoint) {
+    bool isAboveTerrain = true;
+	int	dx, dy;
+	int	incx, incy;
+
+	if (x2 >= x1)
+	{
+		dx = x2 - x1;
+		incx = 1;
+	}
+	else
+	{
+		dx = x1 - x2;
+		incx = -1;
+	}
+
+	if (y2 >= y1)
+	{
+		dy = y2 - y1;
+		incy = 1;
+	}
+	else
+	{
+		dy = y1 - y2;
+		incy = -1;
+	}
+
+	auto x = x1;
+	auto y = y1;
+    int	balance;
+
+	if (dx >= dy)
+	{
+		dy <<= 1;
+		balance = dy - dx;
+		dx *= 2;
+
+		while (x != x2 && isAboveTerrain)
+		{
+			isAboveTerrain &= OnBresenhamPoint(x, y);
+			if (balance >= 0)
+			{
+				y += incy;
+				balance -= dx;
+			}
+			balance += dy;
+			x += incx;
+		} 
+        isAboveTerrain &= OnBresenhamPoint(x, y);
+
+	}
+	else
+	{
+		dx *= 2;
+		balance = dx - dy;
+		dy *= 2;
+
+		while (y != y2  && isAboveTerrain)
+		{
+			isAboveTerrain &= OnBresenhamPoint(x, y);
+			if (balance >= 0)
+			{
+				x += incx;
+				balance -= dy;
+			}
+			balance += dx;
+			y += incy;
+		}
+        isAboveTerrain &= OnBresenhamPoint(x, y);
+	}
+    return isAboveTerrain;
+}
+
+
+// Get the elevation of a single point.
+bool GetElevation(const GDALRasterBandH hBand, const int x, const int y, double& val)
+{
+    return GDALRasterIO(hBand, GF_Read, x, y, 1, 1, &val, 1, 1, GDT_Float64, 0, 0) == CE_None;
+}
+
+// Check a single location is above terrain.
+bool IsAboveTerrain(const GDALRasterBandH hBand, const int x, const int y, const double z)
+{
+    double terrainHeight;
+    if(GetElevation(hBand, x, y, terrainHeight)) {
+        return z > terrainHeight;
+    } else {
+        return false;
+    }
+}
+
+/************************************************************************/
+/*                        GDALIsLineOfSightVisible()                    */
+/************************************************************************/
+
+/**
+ * Check Line of Sight between two points.
+ * Both input coordinates must be within the raster coordinate bounds.
+ *
+ * This algorithm will check line of sight using a 3D bresenham algorithm.
+ * https://www.researchgate.net/publication/2411280_Efficient_Line-of-Sight_Algorithms_for_Real_Terrain_Data
+ *
```

Comment:
A comment could be added to state that the line of sight is computed in raster coordinate space, and thus may not be appropriate for example on datasets referenced against geographic coordinate at high latitudes.

Reply:
Add in cb79c61e07d4e80fc096b6b8840a4f061a9e65c5. 

If there are other approaches to do LOS that would work in these scenarios and we can measure the inaccuracy, please share some references and I can do some research.